# Table of Contents:
1. Exploratory Data Analysis (EDA)
2. Data Preprocessing

# 1. Exploratory Data Analysis (EDA):

In [17]:
import pandas as pd

real_estate_df = pd.read_csv("real_estate_dataset.csv")
real_estate_df.head()

C:\Users\Huyen\AppData\Local\Temp\ipykernel_18792\3545624701.py:3: DtypeWarning: Columns (21) have mixed types. Specify dtype option on import or set low_memory=False.
  real_estate_df = pd.read_csv("real_estate_dataset.csv")


Property_type   Area  Width  Length  Frontage  Number_of_floors  \
0  Nhà biệt thự  196.0    NaN     NaN       NaN               NaN   
1  Nhà biệt thự  196.0    NaN     NaN      10.0               1.0   
2  Nhà biệt thự  282.0    NaN     NaN       NaN               2.0   
3  Nhà biệt thự   80.0    NaN     NaN       NaN               3.0   
4  Nhà biệt thự  337.5    NaN     NaN      15.0               3.0   

   Number_of_bedrooms Number_of_bathrooms  Legal_status    Furniture  ...  \
0                 NaN                 NaN           NaN          NaN  ...   
1                 3.0                   3  Giấy tờ khác          NaN  ...   
2                 4.0                   6      Đã có sổ  Có nội thất  ...   
3                 4.0                   3      Đã có sổ  Có nội thất  ...   
4                 5.0                   6      Đã có sổ  Có nội thất  ...   

  Access_road Posting_date  Expiry_date    Type_of_listing  \
0         NaN   04/11/2023   11/11/2023  Tin VIP Kim Cương   
1        10.0   04/11/2023   11/11/2023       Tin VIP Bạc   
2        16.0   01/11/2023   11/11/2023       Tin VIP Bạc   
3         NaN   01/11/2023   16/11/2023         Tin thường   
4        18.0   31/10/2023   07/11/2023       Tin VIP Bạc   

                       Project_name                  Street          Ward  \
0             Dự án Camellia Garden                     NaN           NaN   
1  Dự án Khu dân cư Camellia Garden                     NaN           NaN   
2                               NaN              Đường số 2   Xã Tân Kiên   
3                 Dự án Lovera Park  Đường Trịnh Quang Nghị  Xã Phong Phú   
4       Dự án Đại Phúc Green Villas         Đường Phạm Hùng  Xã Bình Hưng   

     District        Province Price  
0  Bình Chánh  TP Hồ Chí Minh  18.0  
1  Bình Chánh  TP Hồ Chí Minh  18.0  
2  Bình Chánh  TP Hồ Chí Minh  21.0  
3  Bình Chánh  TP Hồ Chí Minh   6.5  
4  Bình Chánh  TP Hồ Chí Minh  38.0  

[5 rows x 22 columns]

In [18]:
# Check the effect of the null values on the dataset:
for column in real_estate_df.columns:
    unique_values = real_estate_df[column].unique()
    print(f"\033[1m{column}:\033[0m \n {unique_values}")

Property_type: 
 ['Nhà biệt thự' 'Nhà mặt phố' 'Nhà ngõ, hẻm' 'Nhà phố liền kề' 'Chung cư'
 nan]
Area: 
 [196.   282.    80.   ...  62.13 105.95 111.9 ]
Width: 
 [     nan   4.       6.      22.       3.3      3.       3.6      3.8
   7.      12.5      8.5      4.5      4.2     49.       7.8      8.
  11.       5.      11.4      5.5      5.2      2.5      5.82    10.
   3.2      3.5      4.6      4.8     20.       5.4      4.4      5.8
   5.75    12.       5.7      3.4     36.       6.6      3.25     4.3
   4.1      5.16     6.2      3.1      9.5      9.       6.5      3.7
   3.9     13.       3.45     3.71    15.       7.2      8.02    10.5
   7.6     12.15     4.7      6.7      5.3      8.4      5.6      4.05
  14.5      5.1      5.13    16.4     14.       6.1      4.22    26.
  35.       4.02     4.12     3.92    28.      23.      41.      37.
  45.       8.2      9.8      4.25     2.8      4.36    11.5     83.
   7.3     55.       7.5      4.01    11.25     3.85   350.       6.8
  

### Meaning of each field of information:

|No|Attribute name|Meaning 
|---|---|--- 
| 1|Property_type| Loại hình nhà ở, gồm: Nhà biệt thự; Nhà mặt phố; Nhà ngõ, hẻm; Nhà phố liền kề; Chung cư
| 2|Area| Diện tích của bất động sản (đơn vị: m²)
| 3|Width| Chiều rộng của ngôi nhà (đơn vị: m)
| 4|Length| Chiều dài của ngôi nhà (đơn vị: m)
| 5|Frontage| Độ dài mặt tiền của ngôi nhà (đơn vị: m)
| 6|Number_of_floors| Số tầng của nhà mặt đất
| 7|Number_of_bedrooms| Số phòng ngủ của bất động sản
| 8|Number_of_bathrooms| Số phòng tắm của bất động sản
| 9|Legal_status| Tình trạng pháp lý của bất động sản
|10|Furniture| Tình trạng nội thất của bất động sản
|11|House_orientation| Hướng nhà của bất động sản
|12|Balcony_orientation| Hướng ban công của bất động sản
|13|Access_road| Độ dài đường đi vào nhà mặt đất (đơn vị: m)
|14|Posting_date| Ngày đăng tin rao bán bất động sản
|15|Expiry_date| Ngày hết hạn tin rao bán bất động sản
|16|Type_of_listing| Loại tin rao bán bất động sản được đăng lên
|17|Project_name| Tên dự án của bất động sản
|18|Street| Tên Đường/Phố thuộc vị trí của bất động sản
|19|Ward| Tên Phường/Xã/Thị trấn thuộc vị trí của bất động sản
|20|District| Tên Quận/Huyện thuộc vị trí của bất động sản
|21|Province| Tên Tỉnh thành thuộc vị trí của bất động sản
|22|Price| Giá thành của bất động sản (đơn vị: tỷ)

**Observation:**
- Ta thấy cột `Province` chỉ có một giá trị duy nhất là `TP Hồ Chí Minh`, không có ý nghĩa trong dự đoán, vì thế ta sẽ loại bỏ cột này trong phần làm sạch dữ liệu.

In [19]:
rows, cols = real_estate_df.shape
print("rows: ", rows, "\ncols: ", cols)

rows:  57335 
cols:  22


In [20]:
real_estate_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 57335 entries, 0 to 57334
Data columns (total 22 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Property_type        57242 non-null  object 
 1   Area                 57328 non-null  float64
 2   Width                3516 non-null   float64
 3   Length               1823 non-null   float64
 4   Frontage             11478 non-null  float64
 5   Number_of_floors     26218 non-null  float64
 6   Number_of_bedrooms   46226 non-null  float64
 7   Number_of_bathrooms  41008 non-null  object 
 8   Legal_status         35422 non-null  object 
 9   Furniture            23109 non-null  object 
 10  House_orientation    13378 non-null  object 
 11  Balcony_orientation  7409 non-null   object 
 12  Access_road          10602 non-null  float64
 13  Posting_date         57335 non-null  object 
 14  Expiry_date          35501 non-null  object 
 15  Type_of_listing      35562 non-null 

**Observation:**
- Thuộc tính `Number_of_bathrooms` sai kiểu dữ liệu vì có giá trị *Nhiều hơn 6 phòng*. 
-> Tiến hành xử lý giá trị *Nhiều hơn 6 phòng* và đưa kiểu dữ liệu của thuộc tính `Number_of_bathrooms` từ object về float64.

- Ta thấy nhiều thuộc tính có giá trị *Null*.

- Số lượng giá trị *Null* ở một số thuộc tính cũng khá nhiều.
-> Kiểm tra số lượng giá trị *Null* ở mỗi thuộc tính và đưa ra giải pháp.



In [27]:
# Xử lý cột`Number_of_bathrooms:
#real_estate_df['Number_of_bathrooms'].astype('float64')
#real_estate_df['Number_of_bathrooms'].dtype

In [21]:
# Kiểm tra số giá trị Null ở các thuộc tính:
feature_cols = real_estate_df.columns
def display_null(df, feature_cols):
    n_rows = df.shape[0]
    for col in feature_cols:
        null_count = df[col].isnull().sum()
        if null_count > 0:
            print(f"{col} has {null_count*100/n_rows:.2f}% null values.")

display_null(real_estate_df, feature_cols)

Property_type has 0.16% null values.
Area has 0.01% null values.
Width has 93.87% null values.
Length has 96.82% null values.
Frontage has 79.98% null values.
Number_of_floors has 54.27% null values.
Number_of_bedrooms has 19.38% null values.
Number_of_bathrooms has 28.48% null values.
Legal_status has 38.22% null values.
Furniture has 59.69% null values.
House_orientation has 76.67% null values.
Balcony_orientation has 87.08% null values.
Access_road has 81.51% null values.
Expiry_date has 38.08% null values.
Type_of_listing has 37.98% null values.
Project_name has 79.50% null values.
Street has 3.82% null values.
Ward has 0.34% null values.
Price has 0.06% null values.


**Observation:**
- Các thuộc tính như `Width`, `Length`, `Frontage`, `Number_of_floors`, `Furniture`, `House_orientation`, `Balcony_orientation`, `Access_road`, `Project_name` có hơn một nửa là giá trị *Null*.

In [22]:
real_estate_df.describe()

Area        Width       Length      Frontage  Number_of_floors  \
count  57328.000000  3516.000000  1823.000000  11478.000000      26218.000000   
mean     114.704672     5.469146    16.382644      7.021206          3.399954   
std      225.150038     7.520834    10.722797     12.353132          2.828212   
min        7.000000     2.000000     1.000000      1.000000          1.000000   
25%       59.000000     4.000000    12.000000      4.000000          2.000000   
50%       76.000000     4.100000    15.000000      5.000000          3.000000   
75%      112.000000     5.000000    19.000000      8.000000          4.000000   
max    18500.000000   350.000000   222.500000    980.000000        344.000000   

       Number_of_bedrooms   Access_road  
count        46226.000000  10602.000000  
mean             3.471488     11.772364  
std              4.685900     10.676358  
min              1.000000      1.050000  
25%              2.000000      6.000000  
50%              3.000000      8.000000  
75%              4.000000     14.000000  
max            200.000000    322.000000

**Observation:**
- Với cột `Area`, giá trị diện tích lớn nhất là 18500 (m²), nhưng có tới 75% giá trị diện tích là 112 (m²). Việc này chứng tỏ phân bố của dữ liệu này khá lệch (skewed).

- Tương tự với các cột còn lại.

In [23]:
real_estate_df.describe(include='O')

Property_type Number_of_bathrooms Legal_status    Furniture  \
count          57242               41008        35422        23109   
unique             5                   7            3            4   
top         Chung cư                   2     Đã có sổ  Có nội thất   
freq           20466               18154        31444        20575   

       House_orientation Balcony_orientation Posting_date Expiry_date  \
count              13378                7409        57335       35501   
unique                12                  12          180         128   
top           Đông - Nam          Đông - Nam   10/11/2023  23/11/2023   
freq                1630                1314         3989        2052   

       Type_of_listing        Project_name                Street       Ward  \
count            35562               11751                 55145      57141   
unique               4                1281                  5781        186   
top         Tin thường  Dự án Celadon City  Đường Huỳnh Tấn Phát  Phường 12   
freq             34560                 172                   655       2155   

          District        Province  Price  
count        57335           57335  57300  
unique          23               1   3037  
top     TP Thủ Đức  TP Hồ Chí Minh    2.5  
freq         10308           57335    393

**Observation:**
- Với cột `Legal_status`, có *23109* giá trị khác *Null* nhưng lại có đến *20575* giá trị là *Có nội thất*.

- Tương tự với các cột `Type_of_listing`, `Province`.

In [24]:
# Get categorical columns and numerical columns
categorical_cols = [feature for feature in real_estate_df.columns if real_estate_df[feature].dtype == "object"]
numerical_cols = [feature for feature in real_estate_df.columns if real_estate_df[feature].dtype != "object"]
print('Categorical columns: ', categorical_cols)
print('Numerical columns:', numerical_cols)

Categorical columns:  ['Property_type', 'Number_of_bathrooms', 'Legal_status', 'Furniture', 'House_orientation', 'Balcony_orientation', 'Posting_date', 'Expiry_date', 'Type_of_listing', 'Project_name', 'Street', 'Ward', 'District', 'Province', 'Price']
Numerical columns: ['Area', 'Width', 'Length', 'Frontage', 'Number_of_floors', 'Number_of_bedrooms', 'Access_road']


### Data Visualization: